src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,146 @@ namespace scene {
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
 }
 
 SceneNode::SceneNode(MagnumScene& parentNode)
     : Mn::SceneGraph::AbstractFeature3D{*this} {
-  setParent(&parentNode);
+  MagnumObject::setParent(&parentNode);
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
 }
 
-SceneNode& SceneNode::createChild() {
+SceneNode::~SceneNode() {
+  // if the entire scene graph is being deleted, no need to update anything
+  if (SceneGraph::isRootNode(*this)) {
+    return;
+  }
+  // if it is to delete a subtree
+  // then two cases:
+  // 1) current node is the root node (of this subtree).
+  // In case 1, you do not have to do anything since its parent node is nullptr
+  // already!! This is because magnum destroys the nodes in a top-down manner
+  // recursively, root node first, and then its child nodes
+  if (!this->parent()) {
+    return;
+  }
+
+  // 2) current node is NOT the root node.
+  // In case 2, you need update the sensorSuites stored in each ancestor node in
+  // a bottom-up manner.
+
+  // First remove all sensors in nodeSensorSuite from its parent's
+  // nodeSensorSuite if node is leaf node
+  if (getSceneNodeTags() & SceneNodeTag::Leaf) {
+    removeSensorsFromParentNodeSensorSuite();
+  }
+
+  // go bottom to the top, and erase sensors and all subtreeSensors from its
+  // ancestors' subtreeSensorSuites
+  removeSubtreeSensorsFromAncestors();
+}
+
+std::map<std::string, std::reference_wrapper<esp::sensor::Sensor>>&
+SceneNode::getNodeSensors() {
+  return nodeSensorSuite_->getSensors();
+}
+
+std::map<std::string, std::reference_wrapper<esp::sensor::Sensor>>&
+SceneNode::getSubtreeSensors() {
+  return subtreeSensorSuite_->getSensors();
+}
+
+SceneNode& SceneNode::createChild(SceneNodeTags childNodeTags) {
+  CORRADE_ASSERT(
+      !(sceneNodeTags_ & SceneNodeTag::Leaf),
+      "SceneNode::createChild(): Can not create child from leaf node", *this);
   // will set the parent to *this
   SceneNode* node = new SceneNode(*this);
   node->setId(this->getId());
+  node->setSceneNodeTags(childNodeTags);
   return *node;
 }
 
+SceneNode& SceneNode::setParent(SceneNode* newParent) {
+  // Parent can not be leaf node
+  CORRADE_ASSERT(!(newParent->getSceneNodeTags() & SceneNodeTag::Leaf),
+                 "SceneNode::setParent(): New parent node can not be leaf node",
+                 *this);
+
+  // Skip if this is scene or RootNode (which cannot have parent), or if
+  // newParent is already parent
+  if (isScene() || this->parent() == newParent) {
+    return *this;
+  }
+
+  // Skip if this is parent of newParent
```

Comment:
I believe this is being done since this is what Magnum does internally?

Reply:
Yes! This is what magnum is checking internally. Since we update some SensorSuites before calling MagnumObject::setParent(), we have to check again to make sure it's valid to update those SensorSuites.

Reply:
Makes sense.  Can you add a comment?